File: packages/sdk-python/veto/deterministic/regex_safety.py
import re

MAX_PATTERN_LENGTH = 256

_NESTED_QUANTIFIER_ON_GROUP = re.compile(r'[+*}]\s*\)\s*[+*{]')
_ADJACENT_QUANTIFIERS = re.compile(r'[+*}]\s*[+*{]')
_OVERLAPPING_ALTERNATION = re.compile(r'\.\*.*\|.*\.\*')


def is_safe_pattern(pattern: str) -> bool:
    if len(pattern) > MAX_PATTERN_LENGTH:
        return False
    if _NESTED_QUANTIFIER_ON_GROUP.search(pattern):
        return False
    if _ADJACENT_QUANTIFIERS.search(pattern):
        return False
    if _OVERLAPPING_ALTERNATION.search(pattern):
        return False
    return True

File: packages/sdk/src/deterministic/regex-safety.ts
const MAX_PATTERN_LENGTH = 256;
// Detect quantifier inside group followed by quantifier on the group: (a+)+, (a*){2,}, etc.
const NESTED_QUANTIFIER_ON_GROUP = /[+*}]\s*\)\s*[+*{]/;
// Detect directly adjacent quantifiers: a++, a*+, etc.
const ADJACENT_QUANTIFIERS = /[+*}]\s*[+*{]/;
const OVERLAPPING_ALTERNATION = /\.\*.*\|.*\.\*/;

export function isSafePattern(pattern: string): boolean {
  if (pattern.length > MAX_PATTERN_LENGTH) return false;
  if (NESTED_QUANTIFIER_ON_GROUP.test(pattern)) return false;
  if (ADJACENT_QUANTIFIERS.test(pattern)) return false;
  if (OVERLAPPING_ALTERNATION.test(pattern)) return false;
  return true;
}
